### ISSUE TITLE
New update makes reblogging things a mess!

### ISSUE THREAD

**Issue by karuthepanda** [hobinjk/XKit#64]
When reblogging posts it's now impossible to erase any old content. You can't cut posts, you can't erase anything in prior posts even if you wrote it, and it's cluttering the dash and making everything excessively messy. It's also making tumblr unusable for those who RP with GIFs, and it's very frustrating!

**Comment by dlmarquis** [comment]
DEAR SWEET JESUS WHY.

There's a "remove reblogs" option. I could probably write an extension to pull the contents down into the edit box and turn on the "remove" option by default. I'll look at it today and hopefully have it done by tonight.

**Comment by Kikaromi** [comment]
Thanks dlmarquis! You'll make RPers (and probably many more people) very happy.

**Comment by bvtsang** [comment]
Some brief thoughts on @dlmarquis's idea:

You may need to account for the user's text editor. If you pull the contents down into the edit box, you might need to use `$(".editor-richtext").html(quoted_contents)` for rich text and do something else for HTML and Markdown.

Also, for HTML and Markdown, Tumblr seems to wrap the contents of the edit box in `div`s on a line-by-line basis, which might make your implementation annoying at the very least. That might require figuring out Tumblr's styling (such as turning HTML tags a certain color) before parsing.

When the update came out, Tumblr exploded, so I'm looking at this (briefly while at work) to make sure the issue was logged. I won't have the chance to touch this until about 8 hours from now, so either @dlmarquis or @hobinjk might be the better person to fix this.

@karuthepanda, thanks for filing this issue!

----

While I'm at it, we might as well have a formal list of steps to reproduce:

1. Find a post on your dashboard with text in it.
2. Click the reblog button.
3. Expect to be able to edit the reblogged text.
4. Observe that reblogged text is uneditable.

**Comment by dlmarquis** [comment]
I've pretty much got it working for the Rich Text editor. I'll look at the HTML and Markdown editors now. Hopefully they won't cause too many problems.

**Comment by BlackjackOHare** [comment]
https://greasyfork.org/en/scripts/11318-tumblr-reblog-fix/code Somebody made a tampermonkey/greasemonkey script that moves everything into the edit box.

**Comment by anakhasilver** [comment]
That doesn't work on Firefox yet, just gonna point that out.

**Comment by ReddPyramid** [comment]
Grease/Tampermonkey script also conflicts with some X-kit functionality as well, and I'm not a fan of that. X-kit is my go to.

**Comment by BlackjackOHare** [comment]
I linked the code so it could be seen how they did it. :/

**Comment by hobinjk** [comment]
@BlackjackOHare : Thanks for sharing! That is very similar to the approach described by @dlmarquis and @bvtsang. @dlmarquis  is currently eating, otherwise there would already be a patch.

**Comment by ReddPyramid** [comment]
@BlackjackOHare I wasn't trying to put your advice down.  I'm happy to use the help, and i appreciate it.  I apologize if my comment came off harshly.

**Comment by anakhasilver** [comment]
Thanks so much for your work, guys!

**Comment by ReddPyramid** [comment]
And as above, thank you to all of you working to get this resolved!  Lifesavers, every one.

**Comment by BlackjackOHare** [comment]
@ReddPyramid It's alright. ^^ @anakhasilver They updated it to work with firefox. ^^

**Comment by dlmarquis** [comment]
Done! Just waiting on @hobinjk to merge my request.

@bvtsang the extension uses the interface tool XKit.interface.post_window.set_content_html() to paste in the content, which probably needs to be updated to work for HTML and markdown. I'll take a look at it, but you seem to know more about it than me if you want to take a look too.

**Comment by BlackjackOHare** [comment]
I think it's bugged, the xkit extension pastes everything within a blockquote without the extra line at the end outside of a blockquote.

**Comment by dlmarquis** [comment]
@BlackjackOHare Oops, forgot to add the linebreak at the end. Fixed now.

**Comment by bvtsang** [comment]
@dlmarquis I'll take a look at it for HTML and Markdown. Thank you so much!

**Comment by Yunimori** [comment]
Bug in the script. If you go to edit your post -- YOUR post, not the reblogs -- it removes the text you had in there before, so you have to write the whole thing again.

**Comment by dlmarquis** [comment]
@Yunimori should be fixed now.

**Comment by Yunimori** [comment]
@dlmarquis  Ah...that actually seems to have made the entire extension not work. Can't edit reblogs again.

**Comment by dlmarquis** [comment]
@Yunimori Just fixed that (this is what I get for trusting @hobinjk). It pulls the title down as well at the moment but I'm in the middle of fixing that.

**Comment by Yunimori** [comment]
Thanks, seriously. Lemme know when the title pull gets fixed too, and I'll start testing across browsers again. :3

**Comment by hobinjk** [comment]
@Yunimori Just fixed that (this is what I get for trusting @dlmarquis ). It no longer pulls down the title.

**Comment by BlackjackOHare** [comment]
Sweet it's working now. Only weird thing it does is a double linebreak on the new post, but you just have to backspace. ^^

**Comment by therebetterbepie** [comment]
The editable reblogs extension isn't working for me at all. I've installed it, made sure the old xkit was uninstalled and restarted firefox twice. Doesn't work on the dash or any of my drafts. Any advice?

**Comment by hobinjk** [comment]
@therebetterbepie Make sure you're using Editable Reblogs version 1.0.7, clicking the manual update button if necessary. It should be working as long as you do not have the HTML/Markdown editor set as the default.

**Comment by therebetterbepie** [comment]
I just did the manual update and no dice. The rich text editor is my default (I only ever use the HTML editor to do small text) so I don't see why it shouldn't be working but it's not.

**Comment by dlmarquis** [comment]
@therebetterbepie @hobinjk I had the same issue. I had to manually copy the 1.0.7 code into the xkit editor.

**Comment by hobinjk** [comment]
I've tested it on Firefox 39 and Firefox 42, and it appears to work in both. It's possible that Tumblr is in the process of pushing an update which breaks the fix. If you can open editable_reblogs in https://www.tumblr.com/xkit_editor and wrap post_window in `try { original_code_goes_here; } catch(e) { alert(e.message); }` there will be more information available about what is going on.

**Comment by nightpool** [comment]
or just open the console and get the errors? that seems a lot easier to
explain haha.

**Comment by hobinjk** [comment]
@therebetterbepie @dlmarquis I've bumped the version number of editable_reblogs to try to fix that issue.

**Comment by hobinjk** [comment]
nightpool: That actually wasn't working for me in some builds of Firefox due to some bugs with addons and Electrolysis. Of course, YMMV and if that works for you it is definitely easier.

**Comment by nightpool** [comment]
Hmmm. I gave up on e10s when it kept crashing on my macbook and moved to
chrome, but doesn't xkit js inject into the page so it runs in the same
non-addon context tumblr js does? so there should be no process separation?

**Comment by nightpool** [comment]
and I think xkit used to have its own error console you could invoke—does
that also not work or was it removed?

**Comment by nightpool** [comment]
anyway we can take this offline. don't want to spam the people following
the issue.

**Comment by therebetterbepie** [comment]
I tired one more thing before anything else. I used the update all function just to make sure, and that actually worked. Don't know why the manual update didn't work, but at least that's easy.

**Comment by hobinjk** [comment]
On Firefox, XKit now injects into a different context that is bridged by the unsafeWindow wrapper, you can see the details [here](https://github.com/hobinjk/XKit/blob/master/Firefox/index.js).

XKit does still have XKit.console, but it doesn't report uncaught exceptions (yet).

But yes, this is quite a digression from the original issue. I've filed issue #69 to track this.

**Comment by bvtsang** [comment]
As promised, I researched making this work for the HTML and Markdown editors. I know what needs to happen to fix HTML, but I'm a bit stuck.

### HTML

It turns out that Tumblr is using a library called [Ace](http://ace.c9.io/ "Ace - The High Performance Code Editor for the Web") to add syntax highlighting to their post editor. We're fortunate that `window.ace` is accessible to us and that Ace provides some documentation of their API. To insert text into their embedded editor, we need to call `window.ace.edit()` on the DOM node containing their editor, and then call `editor.setValue()`.

If you want a proof of concept of this working, you can do the following:

1. Set your default text editor to HTML.
2. Disable the Editable Reblogs extension.
3. View a post with text and click reblog.
4. In a console, type the following:

```javascript
var editor = ace.edit(document.getElementsByClassName("ace_editor")[0]);
editor.setValue("<p>hello world</p>");
```

To apply this to XKit, here's what vaguely needs to happen. In xkit_patches.js:

```javascript
/**
 * Sets the content of the post window.
 * @param {String} new_content
 */
set_content_html: function(new_content) {
	var editor_type = get_editor_type(); // Magic, to be implemented
	if (editor_type === "rich_text") {
		// stuff goes here
	} else if (editor_type === "html") {
		var editor_div = document.getElementsByClassName("ace_editor")[0];
		// window.ace actually not accessible from current scope!
		var editor = window.ace.edit(editor_div);
		editor.setValue(new_content);
		// Do something about how the inserted text is highlighted
	} else if (editor_type === "markdown") {
		// stuff goes here
	} else {
		// Log an error
	}
},
```

The problem is that `window.ace` is not accessible from that scope, and I can't figure out how to get past that. This is where I was hoping you guys could help out.

### Markdown

I've made less progress on this, but I'm convinced that Tumblr has code to convert HTML to Markdown.

When the dashboard loads, both [this file](https://secure.assets.tumblr.com/client/prod/app/vendor/index.js) and [this file](https://secure.assets.tumblr.com/client/prod/app/context/dashboard/index.js) are retrieved. Pick your favorite Javascript formatter and search for "to-markdown" and "toMarkdown". Specifically, if you search through the first file for "to-markdown", you'll notice the HTML parsing code is very similar to that in [this gist](https://gist.github.com/markandey/4329992 "htmltomarkdown.js"). It's interesting to note the similarities between the files, but this still raises the question of where these methods are defined (such as in closures).

If you do figure out how to use it, then you can simply apply the solution for the HTML editor and allow the Ace editor to handle the rest.

**Comment by hobinjk** [comment]
@bvtsang Great research! It's possible to run scripts in the window scope using the function XKit.tools.add_function. It's a pretty horrible API that I was looking at refactoring, but for now it's the best that exists. There is an okay example of its usage in `soft_refresh`, but the general idea is that whatever function is passed in as its first argument gets turned into a script node and injected into the document. Additionally, data can be retrieved by doing something horrible like `document.getElementById('ext-data').innerHTML = data-you-want-to-retrieve;`

For markdown we probably should just include our own basic html to markdown converter instead of trying to figure out whatever ridiculous nonsense Tumblr has come up with. It's good to know that they appear to do standard things.

**Comment by realityfabric** [comment]
Sometimes when you reblog the post it posts the original and then the edit, even though in the reblog window it only showed the edit

I've only seen one person other than myself with a double post, and they told me that they are using XKit Mobile. I'm using Google Chrome "Version 44.0.2403.125 unknown".

Not sure if I should make a separate bug report for this or not.

**Comment by chooope** [comment]
Not sure if this is the right place to say this or not, but uh-- the editable reblogs extension just isn't showing up for me? At all? I've got new xkit, everything seems to be up to date with it, but when I search for this extension in the gallery it just isn't there. Am I doing something wrong or...?

**Comment by bvtsang** [comment]
@realityfabric Do you mean it reblogs twice, where the first reblog contains the original text and the second reblog contains just the text you appended? Or did you mean something else?

@chooope Is the old XKit uninstalled, disabled, or non-existent (i.e. never installed in the first place)? If so, then could you please file this as a separate issue? Thanks!

**Comment by chooope** [comment]
@bvtsang No, old xkit is there! I did see the post on the blog about old xkit needing to be uninstalled but when I did that xkit just disappeared altogether, and the about:config solution didn't help, so I reinstalled it since everything else works fine, aside from my aforementioned issue...

**Comment by realityfabric** [comment]
@bvtsang No, what is happening is that both the original and edited content appear in one post.

What I think is happening is that the original content is not actually being deleted when it's copied down, but it still looks like it's been deleted while in the reblog box (idk the technical term for it, but the box where you edit posts).

I'd attach an image but I manually fixed it right away, and I haven't seen it happen again, so it might just be a really rare (and easily fixed) bug

**Comment by dlmarquis** [comment]
Currently the extension simulates clicking the remove comments button and then hides the button to bring them back because there wasn't another obvious way to prevent what @realityfabric is describing. This implementation works fine for me, but I've had timing issues with simulated clicks in other extensions (Old Stats) where the click doesn't register. It could be that the same happens in Editable Reblogs, but only occassionally.

**Comment by bvtsang** [comment]
I'm in the middle of working on making this work for Markdown. I've learned that `Xkit.tools.add_function` strips out newlines in its `addt` argument, which affects how Markdown is rendered. I'm reluctant to change `add_function` because I don't know if stripping newlines is preserving some functionality I'm not yet aware of.

Work in progress: https://gfycat.com/GlaringPerfectAssassinbug

**Comment by hobinjk** [comment]
What about running whatever you want addt through JSON.stringify and then JSON.parse? That will properly escape and unescape the newlines for you

**Comment by nightpool** [comment]
I've gotten markdown working, just need to separate it out from the other
changes I have in-tree and PR it.

**Comment by nightpool** [comment]
@bvtsang Did you put a markdown parser in-tree? I think its best in this case to just use the tumblr js function that natively converts html to markdown....

**Comment by nightpool** [comment]
also @hobinjk I think the change I'm going to pr will just change add_function to always use JSON.stringify—it doesn't change anything about the current semantics, (strings still look like strings) and it makes it a lot more intuitive/easier to code with add function.

**Comment by nightpool** [comment]
but this isn't a required part of the change so feel free to challenge that when it comes up or whatever.

**Comment by bvtsang** [comment]
Manually imported https://github.com/domchristie/to-markdown in the same manner as the library used in Elevator, though if you've already gotten it working then I'll leave that to you.

I know Tumblr has a function that converts HTML to Markdown, but I'm not sure what it's called. Minified Javascript isn't easy to look through.

**Comment by nightpool** [comment]
(also its a lot less fragile then just the stripping newlines thing—it protects against having to double escape quotation marks also, for example)

**Comment by nightpool** [comment]
@bvtsang its actually very simple—just a `require("to-markdown").toMarkdown`. 

I sympathize though, Backbone is a fucking pain to step through and it took me at least 20 minutes before I found the actual functions I needed.
